Emit failure resolve event

darkowlzz opened the pull request:
When source reconciliation failure happens, a warning event is emitted but
when resolved, there's no resolved event emitted, leaving the user uninformed
about the status of the operation.

This change checks the object status for failure conditions and
emits a normal event to notify that the failure is resolved on a
successful reconciliation. If a failure is resolved along with a new artifact,
"stored artifact" event is emitted, suppressing the failure resolve message.
New artifact event implies that the reconciliation was successful.

When using notification-controller, this appears as:

![git-repo-fail-final](https://user-images.githubusercontent.com/614105/157894234-2ea92e96-eda0-4e6b-bb2d-b633069580b6.png)

Console logs:

```console
{"level":"error","ts":"2022-03-07T13:57:00.950Z","logger":"controller.gitrepository","msg":"Reconciler error","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"GitRepository","name":"podinfo","namespace":"flux-system","error":"failed to checkout and determine revision: unable to clone 'ssh://git@github.com/stefanprodan/podinfo': Failed to retrieve list of SSH authentication methods: Failed getting response"}
{"level":"info","ts":"2022-03-07T13:57:07.279Z","logger":"controller.gitrepository","msg":"resolved 'GitOperationFailed'","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"GitRepository","name":"podinfo","namespace":"flux-system"}
```

pjbgf reviewed (approved):
This should improve user's perception when facing intermittent issues (i.e. during github clone ops). Nice work @darkowlzz 

LGTM

stefanprodan commented:
That printscreen is confusing, why is there the error twice? Are we sending 2 events now instead of one?

stefanprodan commented:
Maybe it could be something like "resolved <reason>, stored artifact for commit 'Merge pull request #165 from cv65kr/cv65kr-patch-1'"

darkowlzz commented:
> Maybe it could be something like `resolved <reason>, stored artifact for commit <message>`

This event is emitted in `sourceReconcile()` sub-reconciler. By this time, we haven't stored the artifact. We can't say anything about artifact.

darkowlzz commented:
Just for a record, with this, the `ssh could not read data` would appear as:

![git-repo-failure-recovery-2](https://user-images.githubusercontent.com/614105/157062691-cc58bf73-1bf5-4659-a97a-a7979dda5bc2.png)


The simplified messaging could also be:

> resolved 'GitOperationFailed' failure

Not sure how informative we should be.

stefanprodan commented:
I'm not sure we should inject the previous failure in metadata, also do we send a 2nd message after this with the revision? Does this change means people will receive 2 info messages every time SC recovers? For example now SC errors out for each HelmChart at startup, I have 10 charts, so 10 error messages and 10 success messages at every upgrade. Will I receive 30 messages every time after this change?

darkowlzz commented:
> I'm not sure we should inject the previous failure in metadata, also do we send a 2nd message after this with the revision? Does this change means people will receive 2 info messages every time SC recovers? For example now SC errors out for each HelmChart at startup, I have 10 charts, so 10 error messages and 10 success messages at every upgrade. Will I receive 30 messages every time after this change?

Yes, it'll result in double notifications every time there's a failure.
I checked the code, I think we can consolidate everything as one message at the very end, along with the stored artifact message.
Will get back with the new implementation.
Thanks for the feedback.

stefanprodan reviewed (approved):
LGTM

Thanks @darkowlzz

hiddeco commented on controllers/bucket_controller.go:
```diff
@@ -536,28 +570,33 @@ func (r *BucketReconciler) reconcileArtifact(ctx context.Context, obj *sourcev1.
+			Reason: sourcev1.StorageOperationFailedReason,
+		}
+		r.eventLogf(ctx, obj, corev1.EventTypeWarning, sourcev1.StorageOperationFailedReason, e.Err.Error())
+		conditions.MarkTrue(obj, sourcev1.StorageOperationFailedCondition,
```
I feel that with the `StorageOperationFailedCondition` being pretty much equal to `StorageOperationFailedReason`, the arbitrary reason could provide a better summary (in all cases where it is set).

darkowlzz replied:
Addressed by creating granular reasons for all the storage operation failed conditions.

darkowlzz commented:
Summarizing how things changed eventually...
I've updated the PR description with a new image to show what the notifications look like.
Instead of combining recovery and artifact stored events, the recovery event message is suppressed when there's a new artifact. New artifact implies that reconciliation was successful. When there's a recovery without a new artifact, a single notification with the recovery message is sent. Keeping the recovery and new artifact events separate help in watching these events separate with their own counters in K8s native events.
Also, updated all the failures to be associated with some status condition. The storage failures are now associated with the `StorageOperationFailedCondition` with granular reasons for the exact failure which appear in the recovery message.
Updated all the controllers to have these changes.

hiddeco commented:
@stefanprodan please have a look at this and see if you agree with it from a UX perspective. I directed changes this way, as I felt that "polluting" the event data with conditional strings would be non beneficial to native Kubernetes Events, as it for example causes issues with keeping counts of repetitive messages.

By emitting the same event again after "recovering" from a failure, we still produce a signal to a user while allow Kubernetes to keep count.

darkowlzz commented:
Putting this on hold for now. I would like to implement it in another way.
Will extract the API changes into a separate PR.

darkowlzz commented:
#624 is ready, a new simpler implementation. Closing this.
